### ISSUE TITLE
Qemu-nbd is killed during switch root

### ISSUE THREAD

**Issue by lz-lunzi** [systemd/systemd#31690]
### systemd version the issue has been seen with

254

### Used distribution

Fedora 39

### Linux kernel version used

6.7.7-200.fc39.x86_64

### CPU architectures issue was seen on

x86_64

### Component

systemd

### Unexpected behaviour you saw

Fedora kernel: block nbd0: shutting down sockets

### Steps to reproduce the problem

QEMUNBD=$(getarg qemunbd=)

export QEMUNBD
echo  "test:$QEMUNBD"

if [ -n "$QEMUNBD" ]; then
        modprobe  nbd
        modprobe  nvme
        modprobe  ntfs3
        if ismounted "$NEWROOT"; then
		umount "$NEWROOT"
        else
                echo "Directory $NEWROOT is not mounted."
	fi

	
        
      
        ### reset the value of the root variable
        HOSTDEV="${root#block:}"
        echo "blkid: $(blkid)"
        # HOSTDEV="/dev/nvme0n1p1"
        [ -n "$KROOT" ]  ||  root="/dev/nbd0p2"
        # [ -n "$KROOT" ]  &&  root="$KROOT"
        realroot="$root"
        export root
        echo "root：$root"

                        
        ###  auto probe the fs-type of the partition in which vhd-file live and mount it  /host
        mkdir -p /host
        echo "check host：$(ls /host)"

        if [ -z "${HOSTFSTYPE}" ]; then
                HOSTFSTYPE="$(blkid -s TYPE -o value "$HOSTDEV")"
                [ -z "${HOSTFSTYPE}"  -o  "${HOSTFSTYPE}" = "ntfs" ] && HOSTFSTYPE="ntfs-3g"
        fi
        [ "${HOSTFSTYPE}" = "ntfs-3g" ] || modprobe ${HOSTFSTYPE}
        # ln -s "$HOSTDEV" "/run/initramfs/host_device"
         ( exec -a @ntfs-3g ntfs-3g  $HOSTDEV /host ) | (while read l; do warn $l; done)
     
        ### mount the vhd-file on a loop-device
        if [ "${QEMUNBD#/}" != "${QEMUNBD}" ]; then
                
                [ -e /host$QEMUNBD ] || sleep 3
                mkdir -p /var/lock
                ( exec -a @ndb qemu-nbd -c  /dev/nbd0  /host$QEMUNBD   ) | (while read l; do warn $l; done)
                
                udevadm settle

                [ -e "$realroot" ] ||  sleep 3
        fi

        ### mount the realroot / in vhd-file on $NEWROOT
        if [ -z "${KLOOPFSTYPE}" ]; then
                KLOOPFSTYPE="$(blkid -s TYPE -o value "$realroot")"
                [ -z "${KLOOPFSTYPE}" ] && KLOOPFSTYPE="ext4"
        fi
        echo "KLOOPFSTYPE: $KLOOPFSTYPE"
        [ -e "$realroot" ] || sleep 3
        mount -t "ext4" -o rw $realroot $NEWROOT
        if [  "${HOSTHIDDEN}" != "y" ] ; then
                [ -d "${NEWROOT}"/host ] || mkdir -p ${NEWROOT}/host
                echo "NEWROOT: $(ls $NEWROOT)"
                mount -R /host   ${NEWROOT}/host
                
                #  mount -n -o move  /host   ${NEWROOT}/host
        fi
        echo "$(ls ${NEWROOT}/host)"
        echo "$(ls ${NEWROOT})"

fi

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
logs：https://gist.github.com/lz-lunzi/0570a44040270fb676cac4dcbd5c0a96
```

**Comment by DaanDeMeyer** [comment]
https://systemd.io/ROOT_STORAGE_DAEMONS/

**Comment by lz-lunzi** [comment]
@DaanDeMeyer  @Liberasys 
Thank you for your answers

I'm sorry that I'm not an expert in this field. Some data I've found indicate that you can start with @. Actually, there is no change.  ( exec -a @ndb qemu-nbd -c  /dev/nbd0  /host$QEMUNBD   )
Maybe I should learn, but now I want to know how to use bash script to implement
Can you give me some examples

**Comment by lz-lunzi** [comment]
Yes, it worked well before switch_root. It seems to be related to the mechanism of dracut. But I didn't find the way to protect the process during the switch.
References

https://mirrors.edge.kernel.org/pub/linux/utils/boot/dracut/dracut.html#_cleanup_and_switch_root

**Comment by zerodatagravity** [comment]
Here is the patch for systemd (tested on Rocky 9.8 - systemd-0:252-67.el9_8.4.rocky.0.1.x86_64)


```
THE ISSUE:

broadcast_signal() brackets the kill-spree with raw kill(-1, SIGSTOP) / kill(-1, SIGCONT),
bypassing the @ root-storage-daemon marker — kills NBD/FUSE-style daemons at switch-root

THE CODE:

/* src/shared/killall.c, broadcast_signal() */

if (kill(-1, SIGSTOP) < 0 && errno != ESRCH)          /* no ignore_proc() */
        log_warning_errno(errno, "kill(-1, SIGSTOP) failed: %m");

n_children_left = killall(sig, pids, send_sighup);    /* honors '@' */

if (kill(-1, SIGCONT) < 0 && errno != ESRCH)          /* no ignore_proc() */
        log_warning_errno(errno, "kill(-1, SIGCONT) failed: %m");

THE PATCH:

ROOT_STORAGE_DAEMONS.md documents that a daemon which sets argv[0][0] = '@' will be excluded
from the kill spree PID 1 performs before switching from the initrd to the real root. ignore_proc() in
src/shared/killall.c implements that exclusion correctly — but it is only consulted for the signal itself.
The SIGSTOP/SIGCONT pair that brackets it is sent with the raw broadcast form and consults nothing:

--- a/src/shared/killall.c
+++ b/src/shared/killall.c
@@ -266,13 +266,11 @@ int broadcast_signal(int sig, bool wait_for_exit, bool send_sighup, usec_t tim
         assert_se(sigaddset(&mask, SIGCHLD) == 0);
         assert_se(sigprocmask(SIG_BLOCK, &mask, &oldmask) == 0);
 
-        if (kill(-1, SIGSTOP) < 0 && errno != ESRCH)
-                log_warning_errno(errno, "kill(-1, SIGSTOP) failed: %m");
+        (void) killall(SIGSTOP, NULL, false);
 
         n_children_left = killall(sig, pids, send_sighup);
 
-        if (kill(-1, SIGCONT) < 0 && errno != ESRCH)
-                log_warning_errno(errno, "kill(-1, SIGCONT) failed: %m");
+        (void) killall(SIGCONT, NULL, false);
 
         if (wait_for_exit && n_children_left > 0)
                 n_children_left = wait_for_children(pids, &mask, timeout);

So a correctly marked root storage daemon is protected from being terminated and is not protected
from being stopped. For a daemon that holds an interruptible kernel wait open for the lifetime of the
device it serves, a stop is exactly as fatal as a kill — the connection is destroyed during the stop, and
the daemon merely finds out about it after the SIGCONT.

THE PROCEDURE:

dnf download --source systemd

patch it, rebuild it and reinstall it.
```

**Comment by zerodatagravity** [comment]
# Upstream issue FINAL report

Intended for <https://github.com/systemd/systemd/issues>. Related existing
issues: **#31690** ("qemu-nbd is killed during switch root"), **#37700**
(closed via a docs fix), and dracut **#1032**.

---

## Title

`broadcast_signal()` brackets the kill-spree with raw `kill(-1, SIGSTOP)` /
`kill(-1, SIGCONT)`, bypassing the `@` root-storage-daemon marker — kills
NBD/FUSE-style daemons at switch-root

## Summary

`ROOT_STORAGE_DAEMONS.md` documents that a daemon which sets
`argv[0][0] = '@'` will be excluded from the kill spree PID 1 performs
before switching from the initrd to the real root. `ignore_proc()` in
`src/shared/killall.c` implements that exclusion correctly — but it is
only consulted for the signal itself. The `SIGSTOP`/`SIGCONT` pair that
brackets it is sent with the raw broadcast form and consults nothing:

```c
/* src/shared/killall.c, broadcast_signal() */

if (kill(-1, SIGSTOP) < 0 && errno != ESRCH)          /* no ignore_proc() */
        log_warning_errno(errno, "kill(-1, SIGSTOP) failed: %m");

n_children_left = killall(sig, pids, send_sighup);    /* honors '@' */

if (kill(-1, SIGCONT) < 0 && errno != ESRCH)          /* no ignore_proc() */
        log_warning_errno(errno, "kill(-1, SIGCONT) failed: %m");
```

So a correctly marked root storage daemon is protected from being
terminated and is *not* protected from being stopped. For a daemon that
holds an interruptible kernel wait open for the lifetime of the device it
serves, a stop is exactly as fatal as a kill — the connection is destroyed
during the stop, and the daemon merely finds out about it after the
`SIGCONT`.

This makes the documented marker mechanism ineffective for an entire class
of daemons, silently, with no log line from PID 1 indicating that anything
was signalled.

## Affected

Reproduced on systemd 252 (Rocky Linux 9.8, `252-67.el9_8.4.rocky.0.1`), on
**both x86_64 and aarch64** — the latter a Raspberry Pi 4B with an entirely
different kernel and boot chain. Same failure, same fix, so this is neither
architecture- nor platform-specific.

**Still present in systemd 259.** Verified by reading the source of Ubuntu
26.04's `systemd 259.5-0ubuntu3.4` (`apt-get source systemd`,
`src/shared/killall.c`): `broadcast_signal()` there is byte-for-byte
identical to v252, including both raw `kill(-1, …)` calls. Nothing in
`debian/patches/` touches `killall.c`, so that is upstream code shipping
unmodified, not a distro artifact. That is four
releases after `SurviveFinalKillSignal=` was introduced in v255 and
documented as equivalent functionality to the `@` marker — which does not
help here, because it covers the same SIGTERM/SIGKILL step the marker
already covered, not the STOP/CONT bracket.

Worth noting: the function's own comment says the signal goes to all
remaining processes *"if not excluded by `ignore_proc()`"* — a guarantee
that holds for exactly one of the three calls it makes.

`SurviveFinalKillSignal=yes` (v255+) is documented as equivalent
functionality to the `@` marker, so I would expect it to be affected in the
same way, but I have not tested it.

## Reproducer

Boot with the root filesystem served by a userspace daemon that keeps a
thread blocked in the kernel for the connection's lifetime. Ours:

```
qemu-nbd -c /dev/nbd0 -f raw --read-only http://server/rootfs.squashfs
```

started from the initrd inside a unit with `DefaultDependencies=no`,
`IgnoreOnIsolate=yes`, `RefuseManualStop=yes`, and `argv[0]` set to
`@qemu-nbd` via `exec -a`. A squashfs is mounted from `/dev/nbd0` and
overlaid onto `$NEWROOT`. Everything is healthy right up to the pivot.

Any NBD- or FUSE-backed root should reproduce it; the requirement is a
userspace thread parked in an interruptible kernel wait, not NBD
specifically.

### Observed

```
systemd[1]: Switching root.
kernel: block nbd0: shutting down sockets            <- ~38ms later
kernel: I/O error, dev nbd0, sector 3597280 op 0x0:(READ)
kernel: SQUASHFS error: Failed to read block 0x6dc7c1bf: -5
systemd[1]: Failed to resolve /sysroot//mnt: Input/output error
systemd[1]: Failed to switch root, trying to continue: Input/output error
systemd[1]: osee-http-nbd.service: Main process exited, code=exited, status=1/FAILURE
```

Note `switch_root()`'s *first* filesystem operation — `chase_symlinks()`
on `/mnt` — already returns `EIO`. The storage is gone before
`switch_root()` does anything at all, which places the damage in
`broadcast_signal()`, called immediately before it in `do_reexecute()`.

Note also that the marker is verifiably intact at this point
(`/proc/<pid>/cmdline` still begins with `@`), and PID 1 logs no stop, no
kill, and no unit transition for the daemon. The only journal entry is the
`Main process exited` that arrives *after* the damage.

## Evidence

Kernel signal tracepoints, filtered to the daemon's PID, armed from a
dracut `pre-pivot` hook:

```
systemd-1     36.185051: signal_generate: sig=19 comm=qemu-nbd pid=728 grp=1 res=0
qemu-nbd-728  36.185057: signal_deliver:  sig=19 sa_handler=0 sa_flags=0
systemd-1     36.187219: signal_generate: sig=18 comm=qemu-nbd pid=728 grp=1 res=1
qemu-nbd-730  36.189951: signal_generate: sig=15 comm=qemu-nbd pid=728 grp=1 res=0
qemu-nbd-728  36.189959: signal_deliver:  sig=15 sa_handler=55f1c2e07b90 sa_flags=4000004
```

Reading it:

1. PID 1 sends **SIGSTOP** (19) to the marked daemon, group-directed, and
   it is delivered `SIG_DFL`. Every thread stops, including the one parked
   in the kernel. The connection dies here, at `36.185057`.
2. PID 1 sends **SIGCONT** (18) 2.2ms later.
3. 2.7ms after that, the daemon's *own* worker thread sends **SIGTERM**
   (15) to its own process — `qemu-nbd`'s `nbd_client_thread()` does
   `kill(getpid(), SIGTERM)` when its ioctl loop returns an error. This is
   a consequence, not a cause.
4. The main thread handles that SIGTERM and exits.

**No `SIGTERM` from PID 1 appears anywhere** — `killall()` correctly
skipped the marked process for the step that consults `ignore_proc()`. The
marker works. It is simply never consulted for steps 1 and 3.

### Why a stop is fatal here

`drivers/block/nbd.c:nbd_start_device_ioctl()` parks the daemon's thread:

```c
ret = wait_event_interruptible(config->recv_wq,
                               atomic_read(&config->recv_threads) == 0);
if (ret) {
        sock_shutdown(nbd);      /* logs "block nbdX: shutting down sockets" */
        nbd_clear_que(nbd);
}
```

`wait_event_interruptible()` returns non-zero exactly when
`signal_pending()` is true for the waiting thread. A group stop sets
`TIF_SIGPENDING` on every thread, so the `SIGSTOP` pops this wait and the
driver treats it as a disconnect and tears the connection down
unconditionally. Nothing in userspace can recover afterwards, and it makes
no difference whether the signal is caught, ignored by the handler, or
`SA_RESTART` — the teardown has already happened by the time userspace
runs again.

(There is arguably a second bug here, in the NBD driver, which conflates
"this thread received a signal" with "the connection is dead" for a wait
that is by design held open for hours. But the driver behaviour is
long-standing, and `SIGSTOP` cannot be masked from userspace, so the
actionable fix is on the systemd side.)

## Control experiment: the same delivered SIGSTOP is harmless to a daemon
## attached via netlink

To check that the `SIGSTOP` really is the operative cause and not a
coincidence of timing, the same root filesystem was served a second way on
the **same host, same kernel, same unpatched systemd, same pivot**:
`nbdkit` (curl plugin, HTTP→NBD) attached with `nbd-client`'s **netlink**
mode, which hands the connected socket to the kernel and exits, leaving no
userspace thread in `NBD_DO_IT`. Same `@` marker, same unit settings.

```
osee-nbdk: *** NOTE: libsystemd-shared-252.so is UNPATCHED ***
dracut-pre-pivo-706  32.999078: tracing_mark_write: trace-marker start pid=669
systemd-1            33.184673: signal_generate: sig=19 comm=nbdkit pid=669 grp=1 res=0
nbdkit-669           33.184680: signal_deliver:  sig=19 sa_handler=0 sa_flags=0
systemd-1            33.186587: signal_generate: sig=18 comm=nbdkit pid=669 grp=1 res=1
```

The `SIGSTOP` is generated by PID 1 and **delivered**, `SIG_DFL` — the
marked daemon is stopped for 1.9ms and continued, exactly as in the failing
case. But there is no `shutting down sockets`, the device survives intact,
the daemon is alive after the pivot, and the root mounts normally.

Two things follow:

1. The marker is definitively **not** consulted for the STOP/CONT bracket —
   here is a `@`-marked process being stopped by PID 1, observed directly.
   Note also that no `SIGTERM` from PID 1 appears in either capture: the
   marker works perfectly for the one step that consults `ignore_proc()`.
2. The damage is specific to daemons holding an interruptible kernel wait
   open, not to root storage daemons generally. That narrows the blast
   radius but does not shrink the bug: the documented contract is "this
   process is excluded from the kill spree", and stopping it is a violation
   of that contract whether or not a given daemon happens to survive it.

## Proposed fix

`killall()` is a static function already defined earlier in the same file
and already `ignore_proc()`-aware. Use it for the bracket too:

```diff
--- a/src/shared/killall.c
+++ b/src/shared/killall.c
@@ -266,13 +266,11 @@ int broadcast_signal(int sig, bool wait_for_exit, bool send_sighup, usec_t tim
         assert_se(sigaddset(&mask, SIGCHLD) == 0);
         assert_se(sigprocmask(SIG_BLOCK, &mask, &oldmask) == 0);
 
-        if (kill(-1, SIGSTOP) < 0 && errno != ESRCH)
-                log_warning_errno(errno, "kill(-1, SIGSTOP) failed: %m");
+        (void) killall(SIGSTOP, NULL, false);
 
         n_children_left = killall(sig, pids, send_sighup);
 
-        if (kill(-1, SIGCONT) < 0 && errno != ESRCH)
-                log_warning_errno(errno, "kill(-1, SIGCONT) failed: %m");
+        (void) killall(SIGCONT, NULL, false);
 
         if (wait_for_exit && n_children_left > 0)
                 n_children_left = wait_for_children(pids, &mask, timeout);
```

Passing `NULL` for `pids` is already what this code path does for the real
signal (`do_reexecute()` calls `broadcast_signal(SIGTERM, false, …)`, so
`wait_for_exit` is false and `pids` is `NULL` there too).

**Tested:** with this applied, the same boot completes normally, and the
tracepoint capture above comes back with **zero entries** — no signal of
any kind reaches the marked daemon. Without it, the boot fails on every
attempt.

Trade-off worth discussing: `killall()` walks `/proc` and signals each PID
individually, so the stop is no longer atomic across all processes the way
`kill(-1, …)` is, and it costs a `/proc` scan. If that matters for the
shutdown path, an alternative is to keep `kill(-1, …)` but exclude marked
processes explicitly, or to skip the STOP/CONT bracket entirely when
switching root (as opposed to shutting down) — I have no strong opinion on
which shape upstream prefers.

## Also worth considering: make this discoverable

The failure gives PID 1 no log output whatsoever. The daemon is stopped
and continued with nothing recorded, and the first sign of trouble is an
unexplained `EIO` from `chase_symlinks()` inside `switch_root()`. It took
roughly eighteen boot cycles and a kernel tracepoint capture to find, and
`strace` cannot be used to investigate it because `PTRACE_ATTACH` triggers
the identical failure. Even a debug-level line naming what
`broadcast_signal()` is about to do, and to how many processes, would help
the next person a great deal.
